### ISSUE TITLE
Freeze at 'Loading Magic Online' after sign-in

### ISSUE THREAD

**Issue by ubuvoid** [pauleve/docker-mtgo#160]
New issue started presenting after today's MTGO update (2022/09/11). After signing in, application freezes at 'Loading Magic Online' dialog. Logs show the following:

====== MTGO.exe has started.
0324:err:sync:RtlpWaitForCriticalSection section 00D1D070 "?" wait timed out in thread 0324, blocked by 0264, retrying (60 sec)
0324:err:sync:RtlpWaitForCriticalSection section 00D1D070 "?" wait timed out in thread 0324, blocked by 0264, retrying (60 sec)
0324:err:sync:RtlpWaitForCriticalSection section 00D1D070 "?" wait timed out in thread 0324, blocked by 0264, retrying (60 sec)
0324:err:sync:RtlpWaitForCriticalSection section 00D1D070 "?" wait timed out in thread 0324, blocked by 0264, retrying (60 sec)
0324:err:sync:RtlpWaitForCriticalSection section 00D1D070 "?" wait timed out in thread 0324, blocked by 0264, retrying (60 sec)
0324:err:sync:RtlpWaitForCriticalSection section 00D1D070 "?" wait timed out in thread 0324, blocked by 0264, retrying (60 sec)

And so on.

Since it began appearing after today's magic online update I would suspect it's surfaced by a change in MTGO; I don't have enough experience with Wine to speculate about the root cause.

Attaching full log: 
[loading-screen-forever-20220901.txt](https://github.com/pauleve/docker-mtgo/files/9473858/loading-screen-forever-20220901.txt)

**Comment by pauleve** [comment]
I cannot reproduce it here... Maybe give a try to `./run-mtgo panard/mtgo:pr157`
I'll try on another machine tonight.

**Comment by ubuvoid** [comment]
Thanks for the patch. Unfortunately, I get the same behavior on pr157, different address in the log (full logs attached below).

Is there any info I can provide that would help debugging? (Flags I can set to enable verbose logs, etc.)

```
0254:err:combase:RoGetActivationFactory Failed to find library for L"Windows.Foundation.Diagnostics.AsyncCausalityTracer"
====== MTGO.exe has started.
0360:err:sync:RtlpWaitForCriticalSection section 00D1E018 "?" wait timed out in thread 0360, blocked by 0394, retrying (60 sec)
0360:err:sync:RtlpWaitForCriticalSection section 00D1E018 "?" wait timed out in thread 0360, blocked by 0394, retrying (60 sec)
```

[loading-screen-lock-pr157.txt](https://github.com/pauleve/docker-mtgo/files/9480900/loading-screen-lock-pr157.txt)

**Comment by That-Ski-Freak** [comment]
I have the same issue. Frozen on the loading mtgo screen. 
![image](https://user-images.githubusercontent.com/64756991/188237715-4c0c65b2-0c57-4983-85a6-3ce00fb18ea0.png)

**Comment by pauleve** [comment]
You can add the `--debug` option. Also, you may try the emulate desktop option of wine you can set using `--winecfg` (graphics tab)

**Comment by jpark02** [comment]
0370:err:sync:RtlpWaitForCriticalSection section 00D1D040 "?" wait timed out in thread 0370, blocked by 030c, retrying (60 sec)

I'm getting the same thing as well. I've tried the 4-3 image and the 2019 panard images, as well as test, debug, and update. I've also tried panard/mtgo:pr157

**Comment by ubuvoid** [comment]
Reran with --debug enabled. Notable things from the logs:

The following appears in the log upon loading.
```
022c:fixme:reg:RegDisableReflectionKey 0000020C: stub
022c:fixme:reg:RegDisableReflectionKey 00000254: stub
022c:fixme:ole:ftmarshaler_inner_QueryInterface No interface for {00000144-0000-0000-c000-000000000046}
0238:fixme:bcrypt:BCryptGetFipsAlgorithmMode 072EEA8C - semi-stub
022c:fixme:ras:RasEnumConnectionsW (00490BF8,07CFF434,07CFF438),stub!
022c:fixme:ras:RasEnumConnectionsW RAS support is not implemented! Configure program to use LAN connection/winsock instead!
022c:fixme:ras:RasConnectionNotificationW (FFFFFFFF,0000026C,0x00000003),stub!
022c:fixme:nls:LCMapStringEx unsupported lparam 458b80
022c:fixme:ntdll:EtwEventSetInformation (deadbeef, 2, 02732A04, 53) stub
022c:fixme:crypt:SystemFunction041 (00466D4C, 10, 0): stub [RtlDecryptMemory]
022c:fixme:winsock:addrinfo_flags_to_unix unhandled flags 0x20000
0238:fixme:imm:ImeSetActiveContext (0002005E, 1): stub
0238:fixme:imm:ImmReleaseContext (0002006C, 0002005E): stub
022c:fixme:ole:ftmarshaler_inner_QueryInterface No interface for {00000144-0000-0000-c000-000000000046}
022c:fixme:ole:ftmarshaler_inner_QueryInterface No interface for {00000144-0000-0000-c000-000000000046}
```
The message 'No interface for {...]' repeats *a lot*, and after a while we see the following:


```
0264:fixme:ole:ftmarshaler_inner_QueryInterface No interface for {00000144-0000-0000-c000-000000000046}
0264:fixme:ole:ftmarshaler_inner_QueryInterface No interface for {00000144-0000-0000-c000-000000000046}
0264:fixme:ole:ftmarshaler_inner_QueryInterface No interface for {00000144-0000-0000-c000-000000000046}
02e8:fixme:ntdll:EtwEventSetInformation (deadbeef, 2, 53CEF04C, 46) stub
0250:fixme:ole:ftmarshaler_inner_QueryInterface No interface for {00000144-0000-0000-c000-000000000046}
0250:fixme:ole:ftmarshaler_inner_QueryInterface No interface for {00000144-0000-0000-c000-000000000046}
<redundant lines removed>
0250:fixme:ole:ftmarshaler_inner_QueryInterface No interface for {00000144-0000-0000-c000-000000000046}
03b0:err:sync:RtlpWaitForCriticalSection section 00D1DF88 "?" wait timed out in thread 03b0, blocked by 02ec, retrying (60 sec)
0370:fixme:thread:NtQueryInformationThread ThreadIsIoPending info class not supported yet
0378:fixme:thread:NtQueryInformationThread ThreadIsIoPending info class not supported yet
0384:fixme:thread:NtQueryInformationThread ThreadIsIoPending info class not supported yet
```
And so on.

After the first 'RtlpWaitForCriticalSection', we stop seeing 'No interface for {...}'. We only see 'RtlpWaitForCriticalSection' repeat, for as long as I left the app running.

Full logs attached:
[loading-screen-issue160-debug-pr157.txt](https://github.com/pauleve/docker-mtgo/files/9483454/loading-screen-issue160-debug-pr157.txt)

**Comment by dorianbrown** [comment]
Also running into this issue, on two separate computers (both Fedora 36)

**Comment by pauleve** [comment]
I cannot reproduce it on two 2 machines I have...
Have you tried doing a `--reset`?
What is your window manager ?

**Comment by dorianbrown** [comment]
Yes tried that on both machines, still have the issue after `run-mtgo --reset` on both. Using Gnome Shell 4.24 and Docker 20.10.17

**Comment by dorianbrown** [comment]
Also tried enabling the emulate desktop setting in winecfg, but same problem

**Comment by Dominik-Licari** [comment]
I'm also also having the similar issues, attached are the --debug logs from running as both --test and not. When running normally, it sometimes freezes on the loading screen immediately after logging in, and sometimes freezes on the loading magic online page. When running with --test it always makes it to the loading magic online screen. The specific log attached from the non-test run was one freezing on the earlier loading screen.
[logsTest.txt](https://github.com/pauleve/docker-mtgo/files/9486337/logsTest.txt)
[logsRegular.txt](https://github.com/pauleve/docker-mtgo/files/9486338/logsRegular.txt)

**Comment by pauleve** [comment]
I do get the freeze, although not consistently.. it may be related to file i/o.
Could anybody try running wine directly (without docker), installing wine, and following the instructions at https://github.com/pauleve/docker-mtgo/wiki/macOS:-installing-MTGO-using-Wine#first-installation from step 2?

**Comment by Greginator** [comment]
I have it installed using Lutris and can confirm that I am getting the freeze through that. The initial login worked fine but subsequent logins have frozen on the "Loading Magic Online" message.

**Comment by dorianbrown** [comment]
The instructions you referenced in the link aren't available any more for linux, but I installed using Lutris and applied the EULA fix necessary. This installation doesn't seem to suffer from the freezing issue.

**Comment by pauleve** [comment]
I have the same issue using wine directly...

**Comment by pauleve** [comment]
On my side, doing the following helps quite consistently, however, it feels slower:
```
./run-mtgo --shell
taskset -c 0 wine /opt/mtgo/mtgo.exe
```
⚠️ there is still a freeze after login for about 3-5 minutes (an eternity), but then it works.

Also note that doing
```
./run-mtgo --shell
taskset -c 0,1 wine /opt/mtgo/mtgo.exe
```
(enabling 2 CPUs)
seems working as well and feels less laggy.

Using taskset also workaround the issue when using wine directly (outside docker).

If this fixes the issue for most of you, I'll incorporate the change in the docker image, the time that either a new MTGO release is out, or we figure out why there is a threading issue within wine (a nightmare to debug)..

**Comment by jpark02** [comment]
> ./run-mtgo -- --cpuset-cpus 0,1 panard/mtgo:latest 

eventually spat out after 'Loading Magic Online'
> 0258:err:winediag:NtCreateFile Too many open files, ulimit -n probably needs to be increased

This worked to at least load the game but froze when trying to go to Collection.
> ./run-mtgo --shell
> taskset -c 0,1 wine /opt/mtgo/mtgo.exe

**Comment by pauleve** [comment]
> 0258:err:winediag:NtCreateFile Too many open files, ulimit -n probably needs to be increased

That's interesting and shows a potential bug in MTGO.exe which might explain the freezes. Not sure why you don't have this error with the other command though...
You can modify the limits using `ulimit` before launching the Docker image. See https://www.tecmint.com/increase-set-open-file-limits-in-linux/ or https://www.howtogeek.com/805629/too-many-open-files-linux/ for instance.
On my side, I have the following (default) setting:
```$ ./run-mtgo --shell
$ uname -n
1048576
```

**Comment by dorianbrown** [comment]
Adding extra CPUs seems to help for me

**Comment by MultiDragon** [comment]
after the latest update the cpuset fix does not work anymore

**Comment by fugit** [comment]
I was having this problem as well.

The below `./run-mtgo -- --cpuset-cpus 0-3 panard/mtgo:latest` fixed the issue for me. It still hung for a bit but then loaded.

Thanks for all the work on this project.
> ```
> ./run-mtgo -- --cpuset-cpus 0,1 panard/mtgo:latest
> ```
> 
> adventurous users may try enabling more CPUs (e.g., `--cpuset-cpus 0-3` for enabling 4 CPUs)

**Comment by Iceball457** [comment]
Whenever I fully uninstall MTGO and reinstall, the application works fine on the first run. Second and subsequent runs all suffer from hanging after login.
This is true for installs via both docker and lutris, and `./run-mtgo -- --cpuset-cpus 0-3 panard/mtgo:latest` fixes the problem after a short hang (instead of hanging until the OS asks to kill the process)

**Comment by pauleve** [comment]
Since the daybreak transition, I'm no longer experiencing the freeze after login.
Do you confirm?
(you may want to update the [run-mtgo](https://github.com/pauleve/docker-mtgo/raw/master/run-mtgo) script, or use the new [live-run-mtgo](https://github.com/pauleve/docker-mtgo/raw/master/live-run-mtgo)  to always use the latest version of the script)

**Comment by Greginator** [comment]
I can confirm I am still experiencing the freeze after login with the newest script. Was also experiencing it in the version before the newest version.

`./run-mtgo --update
docker pull panard/mtgo:latest
latest: Pulling from panard/mtgo
f57bfd0e3ff4: Already exists 
3ada06f0c2d3: Pull complete 
526c951c3ef5: Pull complete 
458688fb329b: Pull complete 
a5cbc928ab8c: Pull complete 
8f771ced317a: Pull complete 
00fd71eec7f5: Pull complete 
67b50637f32b: Pull complete 
8f6541f713b3: Pull complete 
b6ae2cc266af: Pull complete 
c77006a8b27f: Pull complete 
102dde1ae50a: Pull complete 
690b3fdb9637: Pull complete 
0895c2371a2f: Pull complete 
Digest: sha256:62d643dfd72722ed61f7b709d63097f3c6aff2efd88e2b46e59359906c9d36c6
Status: Downloaded newer image for panard/mtgo:latest
docker.io/panard/mtgo:latest
docker run --privileged --rm -e DISPLAY -v mtgo-data-greg:/home/wine/.wine/host/ -v mtgo-data-greg:/home/wine/.wine/drive_c/users/ -v /tmp/.X11-unix:/tmp/.X11-unix:rw -v /home/greg/.local/share/mtgo/Xauthority:/home/wine/.Xauthority:ro --net=host --ipc=host -e TZ=/usr/share/zoneinfo/Europe/London --name mtgo_running panard/mtgo:latest mtgo
WARNING: The requested image's platform (linux/386) does not match the detected host platform (linux/amd64) and no specific platform was requested
winetricks gdiplus=builtin sound=alsa winegstreamer=disabled wmp=disabled
Executing mkdir -p /home/wine
Using winetricks 20220411-next - sha256sum: d4e94f53f124c5612b26e2eaf7747c31ba081b278d37f2c74d694d9eba1b72ea with wine-7.19 and WINEARCH=win32
Using builtin override for following DLLs: gdiplus
Executing wine regedit C:windowsTemp_builtinoverride-dll.reg
Executing w_do_call sound=alsa
Executing mkdir -p /home/wine
Executing load_sound alsa
Setting sound driver to alsa
Executing wine regedit C:windowsTempset-sound.reg
Using override for following DLLs: winegstreamer
Executing wine regedit C:windowsTemp_disabledoverride-dll.reg
Using override for following DLLs: wmp
Executing wine regedit C:windowsTemp_disabledoverride-dll.reg
wineboot
01a8:err:wineboot:ProcessStartupItems Couldn't get desktop folder.
wine /opt/mtgo/mtgo.exe
01b8:err:vulkan:get_vulkan_driver Wine was built without Vulkan support.
01b0:err:ole:create_server class {20fd4e26-8e0f-4f73-a0e0-f27b8c57be6f} not registered
01b0:err:ole:com_get_class_object no class object {20fd4e26-8e0f-4f73-a0e0-f27b8c57be6f} could be created for context 0x4
01b0:err:ole:create_server class {20fd4e26-8e0f-4f73-a0e0-f27b8c57be6f} not registered
01b0:err:ole:com_get_class_object no class object {20fd4e26-8e0f-4f73-a0e0-f27b8c57be6f} could be created for context 0x4
01b0:err:ole:create_server class {20fd4e26-8e0f-4f73-a0e0-f27b8c57be6f} not registered
01b0:err:ole:com_get_class_object no class object {20fd4e26-8e0f-4f73-a0e0-f27b8c57be6f} could be created for context 0x4
01b0:err:ole:create_server class {20fd4e26-8e0f-4f73-a0e0-f27b8c57be6f} not registered
01b0:err:ole:com_get_class_object no class object {20fd4e26-8e0f-4f73-a0e0-f27b8c57be6f} could be created for context 0x4
01b0:err:ole:create_server class {20fd4e26-8e0f-4f73-a0e0-f27b8c57be6f} not registered
01b0:err:ole:com_get_class_object no class object {20fd4e26-8e0f-4f73-a0e0-f27b8c57be6f} could be created for context 0x4
01b0:err:ole:create_server class {20fd4e26-8e0f-4f73-a0e0-f27b8c57be6f} not registered
01b0:err:ole:com_get_class_object no class object {20fd4e26-8e0f-4f73-a0e0-f27b8c57be6f} could be created for context 0x4
024c:err:actctx:next_xml_elem wrong closing element L"assembly" for L"application"
0254:err:ole:CoGetContextToken apartment not initialised
0268:err:vulkan:get_vulkan_driver Wine was built without Vulkan support.
0268:err:vulkan:get_vulkan_driver Wine was built without Vulkan support.
0268:err:vulkan:get_vulkan_driver Wine was built without Vulkan support.
0268:err:vulkan:get_vulkan_driver Wine was built without Vulkan support.
0268:err:vulkan:get_vulkan_driver Wine was built without Vulkan support.
0268:err:vulkan:get_vulkan_driver Wine was built without Vulkan support.
0268:err:vulkan:get_vulkan_driver Wine was built without Vulkan support.
0268:err:wgl:init_opengl Failed to load libGL: libGL.so.1: cannot open shared object file: No such file or directory
0268:err:wgl:init_opengl OpenGL support is disabled.
0268:err:d3d:wined3d_caps_gl_ctx_create Failed to find a suitable pixel format.
0268:err:d3d:wined3d_adapter_gl_init Failed to get a GL context for adapter 0BF27060.
0268:err:vulkan:get_vulkan_driver Wine was built without Vulkan support.
0268:err:vulkan:get_vulkan_driver Wine was built without Vulkan support.
0268:err:vulkan:get_vulkan_driver Wine was built without Vulkan support.
0268:err:vulkan:get_vulkan_driver Wine was built without Vulkan support.
0268:err:vulkan:get_vulkan_driver Wine was built without Vulkan support.
0268:err:vulkan:get_vulkan_driver Wine was built without Vulkan support.
0268:err:vulkan:get_vulkan_driver Wine was built without Vulkan support.
0268:err:d3d:wined3d_caps_gl_ctx_create Failed to find a suitable pixel format.
0268:err:d3d:wined3d_adapter_gl_init Failed to get a GL context for adapter 00EFD0D0.
ALSA lib confmisc.c:767:(parse_card) cannot find card '0'
ALSA lib conf.c:4745:(_snd_config_evaluate) function snd_func_card_driver returned error: No such file or directory
ALSA lib confmisc.c:392:(snd_func_concat) error evaluating strings
ALSA lib conf.c:4745:(_snd_config_evaluate) function snd_func_concat returned error: No such file or directory
ALSA lib confmisc.c:1246:(snd_func_refer) error evaluating name
ALSA lib conf.c:4745:(_snd_config_evaluate) function snd_func_refer returned error: No such file or directory
ALSA lib conf.c:5233:(snd_config_expand) Evaluate error: No such file or directory
ALSA lib pcm.c:2660:(snd_pcm_open_noupdate) Unknown PCM default
ALSA lib confmisc.c:767:(parse_card) cannot find card '0'
ALSA lib conf.c:4745:(_snd_config_evaluate) function snd_func_card_driver returned error: No such file or directory
ALSA lib confmisc.c:392:(snd_func_concat) error evaluating strings
ALSA lib conf.c:4745:(_snd_config_evaluate) function snd_func_concat returned error: No such file or directory
ALSA lib confmisc.c:1246:(snd_func_refer) error evaluating name
ALSA lib conf.c:4745:(_snd_config_evaluate) function snd_func_refer returned error: No such file or directory
ALSA lib conf.c:5233:(snd_config_expand) Evaluate error: No such file or directory
ALSA lib pcm.c:2660:(snd_pcm_open_noupdate) Unknown PCM default
024c:err:combase:RoGetActivationFactory Failed to find library for L"Windows.Foundation.Diagnostics.AsyncCausalityTracer"
028c:warn:chain:CRYPT_KeyUsageValid keyCertSign not asserted on a CA cert
028c:warn:chain:CRYPT_KeyUsageValid keyCertSign not asserted on a CA cert
028c:err:wininet:HTTP_ParseRfc1123Date unexpected time zone L"UTC"
028c:err:wininet:HTTP_ParseRfc850Date unexpected month format L" Oct 2022 18:00:00 UTC"
028c:err:wininet:HTTP_ParseDate unexpected date format L"Sat, 22 Oct 2022 18:00:00 UTC"
====== MTGO.exe has started.
0444:err:wininet:HTTP_ParseRfc1123Date unexpected time zone L"UTC"
0444:err:wininet:HTTP_ParseRfc850Date unexpected month format L" Oct 2022 18:00:00 UTC"
0444:err:wininet:HTTP_ParseDate unexpected date format L"Sat, 22 Oct 2022 18:00:00 UTC"
04ec:err:sync:RtlpWaitForCriticalSection section 00D1DD18 "?" wait timed out in thread 04ec, blocked by 0608, retrying (60 sec)
04ec:err:sync:RtlpWaitForCriticalSection section 00D1DD18 "?" wait timed out in thread 04ec, blocked by 0608, retrying (60 sec)
04ec:err:sync:RtlpWaitForCriticalSection section 00D1DD18 "?" wait timed out in thread 04ec, blocked by 0608, retrying (60 sec)
04ec:err:sync:RtlpWaitForCriticalSection section 00D1DD18 "?" wait timed out in thread 04ec, blocked by 0608, retrying (60 sec)
04ec:err:sync:RtlpWaitForCriticalSection section 00D1DD18 "?" wait timed out in thread 04ec, blocked by 0608, retrying (60 sec)
04ec:err:sync:RtlpWaitForCriticalSection section 00D1DD18 "?" wait timed out in thread 04ec, blocked by 0608, retrying (60 sec)
`

**Comment by pauleve** [comment]
@Greginator the latest `run-mtgo` script should fix the limit to 4 processes by default, and it does not appear to be the case in your log ; unless you're using it with `./run-mtgo --limit-cpus 0`?
If `./run-mtgo --limit-cpus 4` (the default) still freezes, you may try with a lower number. 
I'll try finding time to debug this issue this week...

**Comment by Greginator** [comment]
Apologies this was an issue my end. Running `./run-mtgo --limit-cpus 4` made me realise I had an outdated version of the run-mtgo script. Having updated that I can confirm it no longer freezes on the login screen after running `./run-mtgo --update`.
